src/NetworkPaymentAdapter.sol

```diff
@@ -0,0 +1,134 @@
+        require(wards[msg.sender] == 1, "Sequencer/not-authorized");
+        _;
+    }
+
+    // --- Data ---
+    VestLike public immutable vest;
+    uint256 public immutable vestId;
+    INetworkTreasury public immutable treasury;
+    DaiJoinLike public immutable daiJoin;
+    DaiLike public immutable dai;
+    address public immutable vow;
+
+    // --- Parameters ---
+    uint256 public bufferMax;
+    uint256 public minimumPayment;
+
+    // --- Tracking ---
+    uint256 public totalSent;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, uint256 data);
+    event TopUp(uint256 bufferSize, uint256 daiBalance, uint256 daiSent, uint256 totalSent);
+
+    // --- Errors ---
+    error InvalidFileParam(bytes32 what);
+    error BufferFull(uint256 bufferSize, uint256 bufferMax);
+    error PendingDaiTooSmall(uint256 pendingDai, uint256 minimumPayment);
+
+    constructor(address _vest, uint256 _vestId, address _treasury, address _daiJoin, address _vow) {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        vest = VestLike(_vest);
+        vestId = _vestId;
+        treasury = INetworkTreasury(_treasury);
+        daiJoin = DaiJoinLike(_daiJoin);
+        dai = DaiLike(daiJoin.dai());
+        vow = _vow;
+    }
+
+    // --- Administration ---
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "bufferMax") {
+            bufferMax = data;
+        } else if (what == "minimumPayment") {
+            minimumPayment = data;
+        } else revert InvalidFileParam(what);
+
+        emit File(what, data);
+    }
+
+    // --- Pay the keeper treasury ---
+    function topUp() external returns (uint256 daiSent) {
+        uint256 bufferSize = treasury.getBufferSize();
+        uint256 pendingDai = vest.unpaid(vestId);
+        uint256 _bufferMax = bufferMax;
+        uint256 _minimumPayment = minimumPayment;
+
+        if (bufferSize + _minimumPayment >= _bufferMax) revert BufferFull(bufferSize, _bufferMax);
+        else if (pendingDai >= _minimumPayment) {
+            vest.vest(vestId);
+            
+            // Send DAI up to the maximum and the rest should go back into the surplus buffer
+            // Use the balance in case someone sends DAI directly to this contract (can be used in emergency)
+            uint256 daiBalance = dai.balanceOf(address(this));
+            if (daiBalance + bufferSize > _bufferMax) {
+                daiSent = _bufferMax - bufferSize;
+
+                // Send the rest back to the surplus buffer
+                daiJoin.join(vow, daiBalance - daiSent);
+            } else {
+                daiSent = daiBalance;
+            }
+            dai.transfer(address(treasury), daiSent);
+            totalSent += daiSent;
+
+            emit TopUp(bufferSize, daiBalance, daiSent, totalSent);
+        } else revert PendingDaiTooSmall(pendingDai, _minimumPayment);
+    }
+
```

Comment:
Does it make sense to add a view function `canTopUp` or something, so keepers can use it to trigger the `topUp` without knowing all details?

Reply:
Yes I'll add this.

Reply:
Added